File: app/src/main/java/com/sunzk/colortest/view/HBSColorSelectorData.java
package com.sunzk.colortest.view;

import androidx.lifecycle.MutableLiveData;

public class HBSColorSelectorData {

	public static final int COLOR_H_MIN = 0;
	public static final int COLOR_H_MAX = 360;

	public static final float COLOR_S_MIN = 0;
	public static final float COLOR_S_MAX = 1.0f;

	public static final float COLOR_B_MIN = 0;
	public static final float COLOR_B_MAX = 1.0f;

	private MutableLiveData<Integer> colorH;
	private MutableLiveData<Float> colorS;
	private MutableLiveData<Float> colorB;

	public HBSColorSelectorData() {
		colorH = new MutableLiveData<>(0);
		colorS = new MutableLiveData<>(0f);
		colorB = new MutableLiveData<>(0f);
	}

	public void setColorH(int h) {
		if (getColorH().getValue() == h) {
			return;
		}
		if (h < COLOR_H_MIN) {
			colorH.postValue(COLOR_H_MIN);
		} else {
			colorH.postValue(Math.min(h, COLOR_H_MAX));
		}
	}

	public MutableLiveData<Integer> getColorH() {
		return colorH;
	}

	public void setColorS(float s) {
		if (getColorS().getValue() == s) {
			return;
		}
		if (s < COLOR_S_MIN) {
			colorS.postValue(COLOR_S_MIN);
		} else {
			colorS.postValue(Math.min(s, COLOR_S_MAX));
		}
	}
	
	public MutableLiveData<Float> getColorS() {
		return colorS;
	}

	public void setColorB(float b) {
		if (getColorB().getValue() == b) {
			return;
		}
		if (b < COLOR_B_MIN) {
			colorB.postValue(COLOR_B_MIN);
		} else {
			colorB.postValue(Math.min(b, COLOR_B_MAX));
		}
	}
	
	public MutableLiveData<Float> getColorB() {
		return colorB;
	}

}

File: app/src/main/java/com/sunzk/colortest/view/HBSColorSelectorData.kt
package com.sunzk.colortest.view

import android.util.Log
import androidx.annotation.IntRange
import com.sunzk.base.expand.emitBy
import com.sunzk.colortest.entity.HSB.Companion.COLOR_B_MAX
import com.sunzk.colortest.entity.HSB.Companion.COLOR_B_MIN
import com.sunzk.colortest.entity.HSB.Companion.COLOR_H_MAX
import com.sunzk.colortest.entity.HSB.Companion.COLOR_H_MIN
import com.sunzk.colortest.entity.HSB.Companion.COLOR_S_MAX
import com.sunzk.colortest.entity.HSB.Companion.COLOR_S_MIN
import kotlinx.coroutines.flow.MutableStateFlow
import kotlin.math.min

class HBSColorSelectorData() {

    companion object {
        private const val TAG: String = "HBSColorSelectorData"
    }
    
    /**
     * from 0 to 360
     */
    val colorH = MutableStateFlow(0)

    /**
     * from 0 to 100
     */
    val colorS = MutableStateFlow(0)
    
    /**
     * from 0 to 100
     */
    val colorB = MutableStateFlow(0)
    
    fun setColorH(@IntRange(from = COLOR_H_MIN.toLong(), to = COLOR_H_MAX.toLong())h: Int) {
        Log.d(TAG, "setColorH: $h")
        val resultColor = 
            if (h < COLOR_H_MIN) {
                COLOR_H_MIN
            } else {
                min(h, COLOR_H_MAX)
            }
        colorH.emitBy(resultColor)
    }

    fun setColorS(@IntRange(from = COLOR_S_MIN.toLong(), to = COLOR_S_MAX.toLong()) s: Int) {
        Log.d(TAG, "setColorS: $s", Throwable())
        val resultColor =
            if (s < COLOR_S_MIN) {
                COLOR_S_MIN
            } else {
                min(s, COLOR_S_MAX)
            }
        colorS.emitBy(resultColor)
    }

    fun setColorB(@IntRange(from = COLOR_B_MIN.toLong(), to = COLOR_B_MAX.toLong())b: Int) {
        Log.d(TAG, "setColorB: $b")
        val resultColor =
            if (b < COLOR_B_MIN) {
                COLOR_B_MIN
            } else {
                min(b, COLOR_B_MAX)
            }
        colorB.emitBy(resultColor)
    }

}